android/src/main/java/com/tailscale/ipn/MainActivity.kt

```diff
@@ -182,12 +186,38 @@ class MainActivity : ComponentActivity() {
     }
   }
 
-  private fun login(url: String) {
-    // (jonathan) TODO: This is functional, but the navigation doesn't quite work
-    // as expected.  There's probably a better built in way to do this.  This will
-    // unblock in dev for the time being though.
-    val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
-    startActivity(browserIntent)
+  private fun login(urlString: String) {
+    lifecycleScope.launch {
+      try {
+        Notifier.state.collect { state ->
+          if (state > Ipn.State.NeedsMachineAuth) {
+            val intent =
+                Intent(applicationContext, MainActivity::class.java).apply {
+                  addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
+                  action = Intent.ACTION_MAIN
+                  addCategory(Intent.CATEGORY_LAUNCHER)
+                }
+            startActivity(intent)
+          }
+        }
+      } catch (e: Exception) {
+        Log.e(TAG, "Login: failed to start MainActivity: $e")
+      }
+    }
+
+    val url = urlString.toUri()
+    try {
+      val customTabsIntent = CustomTabsIntent.Builder().build()
```

Comment:
Cool, I didn't know this was a thing. Seems like a nice solution to not open a bunch of tabs in the user's browser.